Update bundling and pushing a theme.md (#650)
bigcommerce/dev-docs: docs/stencil-docs/deploying-a-theme/bundling-and-pushing.md

<div class="otp" id="no-index">

### On this page
- [Confirm dependencies](#confirm-dependencies)
- [Verify directory and file permissions](#verify-directory-and-file-permissions)
- [Bundling your theme](#bundling-your-theme)
- [Pushing your theme](#pushing-your-theme)
- [Theme quota warning](#theme-quota-warning)
- [Other bundling or upload errors](#other-bundling-or-upload-errors)
- [Resources](#resources)

</div> 

## Confirm dependencies

> If you have customized a theme initially downloaded from the BigCommerce Theme Marketplace: Before you package your theme, make sure your theme directory includes all the dependencies that BigCommerce requires for submission.

Reviewer: ```suggestion
If you have customized a theme originally downloaded from the BigCommerce Theme Marketplace: Before you package your theme, make sure your theme directory includes all the dependencies that BigCommerce requires for submission.

```

<!-- theme: warning -->

### Writable permissions are required
> Without these writable permissions, bundling your theme will fail, blocking its upload to BigCommerce.

</div>
</div>
</div>

## Bundling your theme

> Once you have verified the requirements above, you are ready to process and package your theme for upload to BigCommerce. Stencil CLI provides two options for creating a `.zip` file that contains all of your theme's essentials while excluding redundant components. The options are either only bundling your theme, or bundling and pushing your theme. These options are available depending on how you've [authorized](https://developer.bigcommerce.com/stencil-docs/installing-stencil-cli/troubleshooting-your-setup#stencil-start-errors) your theme:

Reviewer: ```suggestion
Once you have verified the requirements above, you are ready to process and package your theme for upload to BigCommerce. Stencil CLI provides two options for creating a zip file that contains all of your theme's essentials while excluding redundant components. The options are either only bundling your theme, or bundling and pushing your theme. These options are available depending on how you've [authorized](https://developer.bigcommerce.com/stencil-docs/installing-stencil-cli/troubleshooting-your-setup#stencil-start-errors) your theme:

```

### Bundle only

The `stencil bundle` command is available for all themes, whether initialized using OAuth or Basic-Auth tokens. To use it, enter the following on your command line:

`stencil bundle`

The `bundle` command will notify you of its progress and completion.

####  Check/Adjust zipfile's size

> Check the resulting `.zip` file's size before you proceed. The zipped bundle should be only a few megabytes. BigCommerce imposes a hard limit of 50 MB, and any file size approaching that is problematic. If your `.zip` file fits comfortably within the size limit above, jump directly to [Pushing Your Theme](https://developer.bigcommerce.com/stencil-docs/deploying-a-theme/bundling-and-pushing#pushing-your-theme). However, if your `.zip` file approaches or exceeds 50 MB, you must first use one of these procedures to restructure your theme to a manageable size for upload to BigCommerce:

Reviewer: ```suggestion
Check the resulting zip file's size before you proceed. The zipped bundle should be only a few megabytes. BigCommerce imposes a hard limit of 50 MB, and any file size approaching that is problematic. If your zip file fits comfortably within the size limit above, jump directly to [Pushing Your Theme](https://developer.bigcommerce.com/stencil-docs/deploying-a-theme/bundling-and-pushing#pushing-your-theme). However, if your zip file approaches or exceeds 50 MB, you must first use one of these procedures to restructure your theme to a manageable size for upload to BigCommerce:

```
Style guide: [Referring to file types](https://intranet.bigcommerce.com/display/SST/Naming+Guidelines#NamingGuidelines-Referringtofiletypes)

* [Staging a Theme for CDN Delivery](https://support.bigcommerce.com/s/article/Optimizing-Your-Images#imageop-cdn)

### Bundle and push

The `stencil push` command is available only for themes that you have successfully initialized using an OAuth token (with `Themes: modify scope`). This command bundles your theme and uploads it to the associated store, in one continuous process.

For further requirements and usage details, please see the _Command-Line Upload_ in [Pushing Your Theme](/stencil-docs/deploying-a-theme/bundling-and-pushing#pushing-your-theme) below.

For file-size error diagnostics and workarounds, please _Check/Adjust Zipfile's Size_ above.

### Software requirements / resolving lint errors

* Only use the `stencil bundle` or the `stencil push` command to process and package themes for submission. These commands generate `.zip` files that match BigCommerce's expected structure. They also generate metadata required for your theme to function properly.

* Do not create `.zip` files using general-purpose archiving software. The resulting files will trigger errors upon upload to BigCommerce.

> * Do not open a bundled theme `.zip` file to add, delete, rename, or update files. Doing so will make your theme unusable in the production store.

Reviewer: ```suggestion
* Do not open a bundled theme zip file to add, delete, rename, or update files. Doing so will make your theme unusable in the production store.

```

* If bundling your theme triggers multiple lint errors related to the `bundle.js` file, then your theme is missing the `.eslintignore` file. Please retrieve this file from the [Stencil Cornerstone repo](https://github.com/bigcommerce/cornerstone/blob/master/.eslintignore), then re-run `stencil bundle` or `stencil push`.

## Pushing your theme

BigCommerce provides two alternatives for uploading a theme to its associated BigCommerce store. You can perform either a control panel upload, or a command line upload. These options are available depending on how you've authorized your theme:

### Control panel upload (OAuth or Basic Auth)

> Uploading your theme via the BigCommerce control panel's GUI is compatible with any store token (whether OAuth or Basic Auth), with any OAuth token scope, and with all versions of Stencil CLI. However, this option requires that you first use a stencil bundle to package your theme into a .zip file. (Prepare your file according to Bundle only above.)

Author: ```suggestion
Uploading your theme via the BigCommerce control panel's GUI is compatible with any store token (whether OAuth or Basic Auth) and with any OAuth token scope, and with all versions of Stencil CLI. We recommend for users to use OAuth if possible. However, this option requires that you first use a stencil bundle to package your theme into a .zip file. (Prepare your file according to Bundle only above.)

```

> Uploading your theme via the BigCommerce control panel's GUI is compatible with any store token (whether OAuth or Basic Auth), with any OAuth token scope, and with all versions of Stencil CLI. However, this option requires that you first use a stencil bundle to package your theme into a .zip file. (Prepare your file according to Bundle only above.)

Reviewer: ```suggestion
Uploading your theme via the BigCommerce control panel's GUI is compatible with any store token (whether OAuth or Basic Auth), with any OAuth token scope, and with all versions of Stencil CLI. However, this option requires that you first use a stencil bundle to package your theme into a zip file. Prepare your file according to the instructions in the [Bundle only](#bundle-only) section.

```

For the control panel's upload steps, see the [Uploading Custom Themes](https://support.bigcommerce.com/s/article/Stencil-Themes#download-upload) article in our Knowledge Base.

For error codes that you might encounter when uploading a theme – and corresponding workarounds, see [Troubleshooting Theme Uploads](https://developer.bigcommerce.com/stencil-docs/installing-stencil-cli/troubleshooting-your-setup).

### Command line upload (OAuth required)

> The stencil push command allows you to bundle your theme and upload it to the store, with a single terminal command. To run stencil push, you must first:

Reviewer: ```suggestion
The `stencil push` command allows you to bundle your theme and upload it to the store, with a single terminal command. To run `stencil push`, you must first:

```

> Successfully initialize your theme using an OAuth token that you created with the Themes: modify scope.

Reviewer: ```suggestion
Successfully initialize your theme using an OAuth token that you created with the `Themes: modify scope`.

```

![](//s3.amazonaws.com/user-content.stoplight.io/6116/1539055887301 "")

### Successful upload

Upon successful upload, you will receive the prompt, `Would you like to apply your theme to your store at <storehash>? (y/N)`. Any response except `y` or `Y` processes as "No." You can always apply the theme later through the control panel.

### Apply which variation?

> If you chose to apply the newly uploaded theme, you will receive the prompt: `Which variation would you like to apply? (Use arrow keys)`

Reviewer: ```suggestion
If you chose to apply the newly uploaded theme, you will receive the prompt: "Which variation would you like to apply? (Use arrow keys)"

```

![](//s3.amazonaws.com/user-content.stoplight.io/6116/1539055915081 "")

## Theme quota warning

> If you run `stencil push` when your store's `My Themes` section has reached its maximum of 20 themes, you will receive a prompt to select at least one existing theme for deletion.

Reviewer: ```suggestion
If you run `stencil push` when your store's My Themes section has reached its maximum of 20 themes, you will receive a prompt to select at least one existing theme for deletion.

```

Use your arrow keys to move the selection caret to each theme/version that you want to select. Then press the spacebar to select it. (Filled circles will indicate your selected themes/versions.)

> If you are certain of your selections, you can press `Enter` to delete the themes.

Reviewer: ```suggestion
If you are certain of your selections, you can press **Enter** to delete the themes.

```
